jxl/src/frame.rs

```diff
@@ -480,32 +480,49 @@ impl Frame {
     pub fn decode_hf_group(&mut self, group: usize, pass: usize, br: &mut BitReader) -> Result<()> {
         debug!(section_size = br.total_bits_available());
         let lf_global = self.lf_global.as_mut().unwrap();
+        let mut pass_to_pipeline = |chan: usize, g: usize, img: &Image<f32>| {
+            // TODO(veluca): figure out pass count and check group IDs. Also, investigate whether
+            // reusing the `img` buffer would be possible.
+            self.render_pipeline
+                .as_mut()
+                .unwrap()
+                .fill_input_channels(&[chan], g, 1, |rects| rects[0].copy_from(img.as_rect()))
+        };
         if self.header.encoding == Encoding::VarDCT {
             info!("decoding VarDCT");
             let hf_global = self.hf_global.as_mut().unwrap();
             let hf_meta = self.hf_meta.as_mut().unwrap();
-            decode_vardct_group(group, pass, &self.header, lf_global, hf_global, hf_meta, br)?;
+            decode_vardct_group(
+                group,
+                pass,
+                &self.header,
+                lf_global,
+                hf_global,
+                hf_meta,
+                &mut pass_to_pipeline,
+                br,
+            )?;
         }
-
-        let mut pass_to_pipeline = |chan: usize, g: usize, img: &Image<i32>| {
+        let mut convert_and_pass_to_pipeline = |chan: usize, g: usize, img: &Image<i32>| {
             // TODO(veluca): figure out pass count and check group IDs. Also, investigate whether
             // reusing the `img` buffer would be possible.
             self.render_pipeline
                 .as_mut()
                 .unwrap()
-                .fill_input_channels(&[chan], g, 1, |rects| rects[0].copy_from(img.as_rect()))
+                .fill_input_channels(&[chan], g, 1, |rects| {
+                    fill_in_modular_rect(&mut rects[0], &img.as_rect())
```

Comment:
Ideally, the idea here would be to pass `i32` channels to the pipeline, and have the pipeline convert to f32 where needed.
That functionality should be supported, but the stage to do this might not be present yet.

Reply:
We only have a dummy render pipeline now that just stitches the groups together. I added a TODO to move this to the render pipeline.